# 
# Submitted:
#
# Robert Schwebel, 2003-11-19
# Benedikt Spranger, 2004-02-06
#
# Error: 
#
# (echo '#include <sysdep-cancel.h>'; \
#  echo 'PSEUDO (__libc_pause, pause, 0)'; \
#  echo ' ret'; \
#  echo 'PSEUDO_END(__libc_pause)'; \
#  echo 'libc_hidden_def (__libc_pause)'; \
#  echo 'weak_alias (__libc_pause, pause)'; \
#  echo 'libc_hidden_weak (pause)'; \
# ) | arm-unknown-linux-gnu-gcc -c  [...] 
# <stdin>: Assembler messages:
# <stdin>:2: Error: garbage following instruction -- `ldr lr,[sp],#4 ldr ip,=__libc_multiple_threads' 
# ...
# memory fault while connect()
#
# Description: 
#
# http://sources.redhat.com/ml/libc-alpha/2003-05/msg00097.html
# 
# see also: 
#
# http://mail.gnu.org/archive/html/bug-glibc/2003-05/msg00075.html
#
# State:
# 
# should be solved in glibc CVS
#

--- glibc-2.3.2/linuxthreads/sysdeps/unix/sysv/linux/arm/sysdep-cancel.h.orig	2003-01-12 10:13:57.000000000 +0100
+++ glibc-2.3.2/linuxthreads/sysdeps/unix/sysv/linux/arm/sysdep-cancel.h	2004-02-06 17:55:20.000000000 +0100
@@ -24,6 +24,15 @@
 
 #if !defined NOT_IN_libc || defined IS_IN_libpthread
 
+/* We push lr onto the stack, so we have to use ldmib instead of ldmia
+   to find the saved arguments.  */
+#undef DOARGS_5
+#undef DOARGS_6
+#undef DOARGS_7
+#define DOARGS_5 str r4, [sp, $-4]!; ldr r4, [sp, $8];
+#define DOARGS_6 mov ip, sp; stmfd sp!, {r4, r5}; ldmib ip, {r4, r5};
+#define DOARGS_7 mov ip, sp; stmfd sp!, {r4, r5, r6}; ldmib ip, {r4, r5, r6};
+
 # undef PSEUDO_RET
 # define PSEUDO_RET						        \
     ldrcc pc, [sp], $4;						        \
@@ -34,7 +43,7 @@
 # define PSEUDO(name, syscall_name, args)				\
   .section ".text";							\
     PSEUDO_PROLOGUE;							\
-  ENTRY (name)								\
+    ENTRY (name);							\
     SINGLE_THREAD_P_INT;						\
     bne .Lpseudo_cancel;						\
     DO_CALL (syscall_name, args);					\
@@ -110,7 +119,7 @@
   ldr reg, 2b;								\
 3:									\
   add ip, pc, ip;							\
-  ldr ip, [ip, lr];							\
+  ldr ip, [ip, reg];							\
   teq ip, #0;
 #   define SINGLE_THREAD_P_INT						\
   str lr, [sp, $-4]!;							\
